### PR TITLE
[FLINK-14918] [flink-end-to-end-perf-tests] Add performance e2e test module and scripts

### PR DESCRIPTION
**Pull request by Li-Aihua**
## What is the purpose of the change

*Add performance e2e test module and scripts which methoned in .FLINK-14918*

## Brief change log

*This change adds performance e2e test module and scripts to run end-2-end performance test and collects metrics *

## Verifying this change

*Manually verified that  can run correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: tools/jenkins/restapi_common.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import urllib2
+import json
+
+
+def execute_get(url):
+    result_url = ""
+    try:
+        result_url = urllib2.urlopen(url, timeout=300).read()
+    except urllib2.URLError, e:
+        print e.reason
```
Comment (klion26) [inline tools/jenkins/restapi_common.py]: maybe we should use log instead of print

File: tools/jenkins/restapi_common.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+#
+# This common file writes the functions of getting qps from flink restful api
+#
+# Note by AiHua Li:
+
+import urllib2
+import json
+
+
+def execute_get(url):
```
Comment (klion26) [inline tools/jenkins/restapi_common.py]: do you mind adding some doc for function?

File: tools/jenkins/restapi_common.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        print e.reason
+    return result_url
+
+
+def get_source_node(plan):
+    source_nodes = []
+    nodes = plan["nodes"] if "nodes" in plan else []
+    for node in nodes:
+        inputs = node["inputs"] if "inputs" in node else []
+        if len(inputs) == 0:
```
Comment (klion26) [inline tools/jenkins/restapi_common.py]: maybe here we want to append to `source_nodes` when `len(inputs)` is _not_ zero?

Comment (klion26) [inline tools/jenkins/restapi_common.py]: maybe here we want to append to `source_nodes` when `len(inputs)` is _not_ zero?

Comment (klion26) [reply tools/jenkins/restapi_common.py]: do we need to add ut for functions?

File: tools/jenkins/restapi_common.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            source_nodes.append(node["id"])
+    return source_nodes
+
+
+def get_avg_qps_by_restful_interface(am_seserver_dddress, job_id):
+    url = "http://%s/jobs/%s" % (am_seserver_dddress, job_id)
+    result = execute_get(url)
```
Comment (klion26) [inline tools/jenkins/restapi_common.py]: what will we do after `execute_get(url)` throws exception?

File: tools/jenkins/restapi_common.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    result = json.loads(result)
```
Comment (klion26) [inline tools/jenkins/restapi_common.py]: will this throw an exception if the result is invalid? how do we want  to do in such case

File: tools/jenkins/restapi_common.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    vertices = result["vertices"] if "vertices" in result else ""
```
Comment (klion26) [inline tools/jenkins/restapi_common.py]: can we use `get(key, [default])` here? and other places

File: tools/jenkins/restapi_common.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    plan = result["plan"] if "plan" in result else ""
+    source_nodes = get_source_node(plan)
+    for vertice in vertices:
+        id = vertice["id"] if "id" in vertice else ""
+        if id in source_nodes:
+            url="http://%s/jobs/%s/vertices/%s/subtasks/metrics?agg=avg" % (am_seserver_dddress, job_id, id)
+            keyresult=execute_get(url)
```
Comment (klion26) [inline tools/jenkins/restapi_common.py]: blank around `=`

File: tools/jenkins/restapi_common.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                metrics_name=key["id"]
+                if metrics_name.endswith("numBuffersOutPerSecond"):
+                    url="http://%s/jobs/%s/vertices/%s/subtasks/metrics?get=%s" % (am_seserver_dddress, job_id, id,
+                                                                                  metrics_name)
+                    value_result = execute_get(url)
+                    value_result = json.loads(value_result)
+                    for value in value_result:
+                        tps = value["avg"]
+                        if tps > 0:
+                            totaltps = totaltps + tps
```
Comment (klion26) [inline tools/jenkins/restapi_common.py]: maybe we need better to init `totaltops` outside?

File: tools/jenkins/run_case.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+import time
+
+from utils import run_command
+from restapi_common import get_avg_qps_by_restful_interface
+
+
+def start_server(flink_home):
+    cmd = "bash %s/bin/start_yarn.sh" % flink_home
+    status, output = run_command(cmd)
+    if status ==0:
```
Comment (klion26) [inline tools/jenkins/run_case.py]: blank after `==`

File: tools/jenkins/run_case.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                qps = get_avg_qps_by_restful_interface(am_seserver_dddress)
+                total_qps.append(qps)
+                time.sleep(wait_minute)
+        avg_qps = get_avg(total_qps)
+        print "The avg qps of %'s  is %s" % (scenario_name, avg_qps)
+
+
+
+if __name__ == "__main__":
+    am_seserver_dddress = sys.argv[1]
```
Comment (klion26) [inline tools/jenkins/run_case.py]: maybe we should check how maybe arguments passed in first

File: tools/jenkins/utils.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def run_command(cmd, timeout=12000):
+    script = '''#!/bin/sh
+    %s
+    '''%cmd
+    script_file = tempfile.NamedTemporaryFile('wt')
+    script_file.write(script)
+    script_file.flush()
+    outputFile = "/tmp/%s"%uuid.uuid1()
```
Comment (klion26) [inline tools/jenkins/utils.py]:
```suggestion
    outputFile = "/tmp/%s" % uuid.uuid1()
```

File: tools/jenkins/utils.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    stderrFile = "/tmp/%s"%uuid.uuid1()
```
Comment (klion26) [inline tools/jenkins/utils.py]:
```suggestion
    stderrFile = "/tmp/%s" % uuid.uuid1()
```

**Comment by Li-Aihua** [comment]
> @Li-Aihua thanks for your contribution, left some comments.

Thanks klion26 for you suggestion, I fix all places which you said, please review again.

File: tools/jenkins/utils.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+################################################################################
+
+import subprocess
+import traceback
+from logger import logger
+from threading import Timer
+
+
+def timeout_callback(p):
+    print 'exe time out call back'
```
Comment (klion26) [inline tools/jenkins/utils.py]: we can use logger here

File: tools/jenkins/run_case.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+            return 1
+        for inter_index in range(0, inter_nums):
+            cmd = "bash %s/bin/flink run %s" % (flink_home, scenario)
+            status, output = run_command(cmd)
+            if status == 0:
+                qps = get_avg_qps_by_restful_interface(am_seserver_dddress)
+                total_qps.append(qps)
+                time.sleep(wait_minute)
+        avg_qps = get_avg(total_qps)
+        logger.info("The avg qps of %'s  is %s" % (scenario_name, avg_qps))
```
Comment (klion26) [inline tools/jenkins/run_case.py]: Why do we need to add `'` between `%` and `s` here

File: tools/jenkins/run_case.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        scenarios.append(cmd)
+        line = scenario_file.readline()
+    scenario_file.close()
+    return 0, scenarios, scenario_names
+
+
+def get_avg(values):
+    sumvalues = 0
+    value_num = 0
+    for value in values:
```
Comment (klion26) [inline tools/jenkins/run_case.py]: here we can test whether len(values) is 0, if it's zero, we will return 0 directly, other than, we can use `sum(values) * 1.0 / len(values)` to calculate the average value of the passed in list

File: tools/jenkins/run_case.py [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+    else:
+        return 1
+
+
+def end_server(flink_home):
+    cmd = "bash %s/bin/stop_yarn.sh" % flink_home
+    run_command(cmd)
+
+
+def get_scenarios(scenario_file_name, test_jar):
```
Comment (klion26) [inline tools/jenkins/run_case.py]: maybe we can add some documentation for the functions, what does it do, what do the parameters mean, and what result will return

**Comment by Li-Aihua** [comment]
Thanks klion26 , I fix all places which you said, please check again.thanks

File: tools/jenkins/utils.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    try:
+        p.kill()
+    except Exception as error:
+        logger.error(error)
```
Comment (zentol) [inline tools/jenkins/utils.py]: Needs some additional context, like "Failed to kill process spawned by command <cmd>"

File: tools/jenkins/run_case.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+
+if __name__ == "__main__":
+    if len(sys.argv)<3:
+        logger.error("The param's number must be larger than 3")
+        sys.exit(1)
+    am_seserver_dddress = sys.argv[1]
+    scenario_file = sys.argv[2]
+    flink_home = sys.argv[3]
+    if len(sys.argv) >= 4:
```
Comment (zentol) [inline tools/jenkins/run_case.py]: this looks like a rather brittle approach; why aren't we using `argparse` for this?

File: tools/jenkins/run_case.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+                time.sleep(wait_minute)
+        avg_qps = get_avg(total_qps)
+        logger.info("The avg qps of %s's  is %s" % (scenario_name, avg_qps))
+
+
+if __name__ == "__main__":
+    if len(sys.argv)<3:
+        logger.error("The param's number must be larger than 3")
+        sys.exit(1)
+    am_seserver_dddress = sys.argv[1]
```
Comment (zentol) [inline tools/jenkins/run_case.py]: There's a type in the name and I have no idea what `am_seserver_dddress` even is

**Pull request closed without merging**